pycaw/magic.py

```diff
@@ -0,0 +1,812 @@
+        return new_mute
+
+    def step_volume(self, step=0.1):
+        current = self.volume
+        new = max(0, min(1, current + step))
+        if new != current:
+            self.volume = new
+            return new
+        return None
+
+
+class MagicSession():
+    """
+    When activated and passed via
+    MagicManager.magic_session(inherited_class_of_MagicSession)
+    will be created for each new and current session.
+    Dict of all iid -> MagicSession:
+        MagicManager.magic_sessions
+    """
+    guid = pointer(GUID('{34482A3D-37DD-40E3-BB37-63A16036C87C}'))
+
+    def __init__(self,
+                 volume_callback=None,
+                 mute_callback=None,
+                 state_callback=None):
+
+        self.magic_root_session = self._passed_magic_root_session
+        self._passed_magic_root_session = None
+
+        # callbacks
+        self.volume_callback = volume_callback
+        self.mute_callback = mute_callback
+        self.state_callback = state_callback
+
+        # tells the magic_root_session to create a connection
+        # for callbacks:
+        self.magic_root_session.use_magic_session(self)
+
+        log.info(str(self))
+
+    @classmethod
+    def initialize(cls, new_magic_root_session, *args, **kwargs):
+        """
+        is called by the MagicManager.
+
+        Every new MagicSession needs to have a magic_root_session.
+
+        instead of passing it visible as argument trough the inherited class
+        back to the super().__init(new_magic_root_session) -
+
+        the new_magic_root_session it is transferred by setting
+        a class attribute and picking it up in the MagicSession.__init__()
+
+        class PassingViaArg(MagicSession):
+            def __init__(self, new_magic_root_session, *args, **kwargs):
+                super().__init__(new_magic_root_session)
+
+        class PassingViaClassAttr(MagicSession):
+            def __init__(self, *args, **kwargs):
+                super().__init__()
+        """
+        cls._passed_magic_root_session = new_magic_root_session
+
+        magic_session = cls(*args, **kwargs)
+        return magic_session
+
+    def __str__(self):
+        return (f"<{self.__class__.__name__} "
+                f"app_exe='{self.magic_root_session.app_exe}'/>")
+
+    # easy control:
+    # TODO: handle incorrect input or raise exception.
+    # also handle failing com calls
+    # (failing in terms of the retrieved value is not 'S_OK')
+    @property
+    def state(self):
+        return self.magic_root_session.state
+
+    @property
+    def volume(self):
+        return self.magic_root_session.volume
+
+    @volume.setter
+    def volume(self, volume):
+        self.magic_root_session._sav.SetMasterVolume(volume, self.guid)
+
+    @property
+    def mute(self):
+        return self.magic_root_session.mute
+
+    @mute.setter
+    def mute(self, mute):
+        self.magic_root_session._sav.SetMute(mute, self.guid)
+
+    def toggle_mute(self):
```

Comment:
Since `toggle_mute` and `step_volume` at least are used in two difference placed, maybe we could defined it as a function somewhere and call it from the method of both `MagicSession` and `MagicApp` what do you think?

Reply:
Let me introduce you to yet another parrent class ;)